selfdrive/pandad/panda/state_manager.py

```diff
@@ -0,0 +1,126 @@
+import os
+from cereal import car, log
+import cereal.messaging as messaging
+from panda.python import PANDA_BUS_CNT
+
+SPOOFING_STARTED = os.getenv("STARTED") == "1"
+
+LEC_ERROR_CODE = {
+  "No error": 0, "Stuff error": 1, "Form error": 2, "AckError": 3,
+  "Bit1Error": 4, "Bit0Error": 5, "CRCError": 6, "NoChange": 7,
+}
+
+class PandaStateManager:
+  def __init__(self, pandas, hw_types, lock):
+    self.pandas = pandas
+    self.lock = lock
+    self.hw_types = hw_types
+
+    self.is_comma_three_red = (
+      len(pandas) == 2 and
+      self.hw_types[0] == log.PandaState.PandaType.dos and
+      self.hw_types[1] == log.PandaState.PandaType.redPanda
+    )
+
+  def process(self, engaged, pm) -> bool:
+    msg = messaging.new_message('pandaStates', len(self.pandas))
+    msg.valid = True
+    panda_states = msg.pandaStates
+    ignition = False
+
+    for i, panda in enumerate(self.pandas):
+      with self.lock:
+        health = panda.health() or {}
+
+      if SPOOFING_STARTED:
+        health['ignition_line'] = 1
+      elif self.is_comma_three_red and self.hw_types[i] == log.PandaState.PandaType.dos:
+        health['ignition_line'] = 0
+
+      ignition |= bool(health['ignition_line']) or bool(health['ignition_can'])
+      ps = panda_states[i]
+      self._fill_state(ps, self.hw_types[i], health)
+
+      # Fill can state
+      for j in range(PANDA_BUS_CNT):
+        with self.lock:
+          can_health = panda.can_health(j)
+        can_state = ps.init(f'canState{j}')
+        self._fill_can_state(can_state, can_health)
+
+      # Set faults
+      fault_bits = int(health.get('faults', 0))
+      faults_list = [f for f in range(log.PandaState.FaultType.relayMalfunction,
+                                      log.PandaState.FaultType.heartbeatLoopWatchdog + 1)
+                    if fault_bits & (1 << f)]
+      faults = ps.init('faults', len(faults_list))
+      for idx, fault in enumerate(faults_list):
+        faults[idx] = fault
+
+    with self.lock:
+      for panda, ps in zip(self.pandas, panda_states):
+        if ps.safetyModel == car.CarParams.SafetyModel.silent or (
+            not ignition and ps.safetyModel != car.CarParams.SafetyModel.noOutput):
+          panda.set_safety_mode(car.CarParams.SafetyModel.noOutput)
+
+        if ps.powerSaveEnabled != (not ignition):
+          panda.set_power_save(not ignition)
+
+        panda.send_heartbeat(engaged)
+
+    pm.send("pandaStates", msg)
+    return ignition
+
+  def _fill_state(self, ps, hw_type, health):
+    ps.voltage = health['voltage']
```

Comment:
I propose we refactor the panda health reponse to output a format compatible for dict init on the capnp message so we can do:
```
msg = messaging.new_message('pandaStates', 0)
msg.pandaStates[0] = panda.health()
```